File: neural-engine/src/api/sdk/javascript/src/exceptions.ts
/**
 * NeuraScale SDK Exceptions
 */

export class NeuraScaleError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public response?: any,
  ) {
    super(message);
    this.name = "NeuraScaleError";
  }
}

export class AuthenticationError extends NeuraScaleError {
  constructor(message: string = "Authentication failed") {
    super(message, 401);
    this.name = "AuthenticationError";
  }
}

export class AuthorizationError extends NeuraScaleError {
  constructor(message: string = "Insufficient permissions") {
    super(message, 403);
    this.name = "AuthorizationError";
  }
}

export class NotFoundError extends NeuraScaleError {
  constructor(message: string = "Resource not found") {
    super(message, 404);
    this.name = "NotFoundError";
  }
}

export class ValidationError extends NeuraScaleError {
  constructor(
    message: string,
    public details?: any,
  ) {
    super(message, 422);
    this.name = "ValidationError";
  }
}

export class RateLimitError extends NeuraScaleError {
  constructor(
    message: string = "Rate limit exceeded",
    public retryAfter: number = 60,
  ) {
    super(message, 429);
    this.name = "RateLimitError";
  }
}

export class ServerError extends NeuraScaleError {
  constructor(message: string = "Internal server error") {
    super(message, 500);
    this.name = "ServerError";
  }
}

export class ConnectionError extends NeuraScaleError {
  constructor(message: string = "Connection failed") {
    super(message);
    this.name = "ConnectionError";
  }
}

export class TimeoutError extends NeuraScaleError {
  constructor(message: string = "Request timeout") {
    super(message);
    this.name = "TimeoutError";
  }
}

File: neural-engine/src/api/sdk/python/neurascale/exceptions.py
"""NeuraScale SDK exceptions."""

from typing import Optional, Dict, Any


class NeuraScaleError(Exception):
    """Base exception for NeuraScale SDK."""

    def __init__(
        self,
        message: str,
        status_code: Optional[int] = None,
        response: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.response = response


class AuthenticationError(NeuraScaleError):
    """Authentication failed."""

    def __init__(self, message: str = "Authentication failed"):
        super().__init__(message, status_code=401)


class AuthorizationError(NeuraScaleError):
    """Authorization failed."""

    def __init__(self, message: str = "Insufficient permissions"):
        super().__init__(message, status_code=403)


class NotFoundError(NeuraScaleError):
    """Resource not found."""

    def __init__(self, message: str = "Resource not found"):
        super().__init__(message, status_code=404)


class ValidationError(NeuraScaleError):
    """Validation error."""

    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, status_code=422)
        self.details = details


class RateLimitError(NeuraScaleError):
    """Rate limit exceeded."""

    def __init__(self, message: str = "Rate limit exceeded", retry_after: int = 60):
        super().__init__(message, status_code=429)
        self.retry_after = retry_after


class ServerError(NeuraScaleError):
    """Server error."""

    def __init__(self, message: str = "Internal server error"):
        super().__init__(message, status_code=500)


class ConnectionError(NeuraScaleError):
    """Connection error."""

    def __init__(self, message: str = "Connection failed"):
        super().__init__(message)


class TimeoutError(NeuraScaleError):
    """Request timeout."""

    def __init__(self, message: str = "Request timeout"):
        super().__init__(message)
